pkg/agent/openflow/pipeline.go

```diff
@@ -1372,34 +1427,40 @@ func (c *client) conjunctionActionFlow(conjunctionID uint32, tableID binding.Tab
 	return flows
 }
 
-// conjunctionActionDropFlow generates the flow to mark the packet to be dropped if policyRuleConjunction ID is matched.
+// conjunctionActionDenyFlow generates the flow to mark the packet to be denied
+// (dropped or rejected) if policyRuleConjunction ID is matched.
 // Any matched flow will be dropped in corresponding metric tables.
-func (c *client) conjunctionActionDropFlow(conjunctionID uint32, tableID binding.TableIDType, priority *uint16, enableLogging bool) binding.Flow {
+func (c *client) conjunctionActionDenyFlow(conjunctionID uint32, tableID binding.TableIDType, priority *uint16, disposition uint32, enableLogging bool) binding.Flow {
 	ofPriority := *priority
 	metricTableID := IngressMetricTable
 	if _, ok := egressTables[tableID]; ok {
 		metricTableID = EgressMetricTable
 	}
-	// We do not drop the packet immediately but send the packet to the metric table to update the rule metrics.
+
+	flowBuilder := c.pipeline[tableID].BuildFlow(ofPriority).
+		MatchConjID(conjunctionID).
+		Action().LoadRegRange(int(CNPDenyConjIDReg), conjunctionID, binding.Range{0, 31}).
+		Action().LoadRegRange(int(marksReg), cnpDenyMark, cnpDenyMarkRange)
+
 	if enableLogging {
-		return c.pipeline[tableID].BuildFlow(ofPriority).
-			MatchConjID(conjunctionID).
-			Action().LoadRegRange(int(CNPDropConjunctionIDReg), conjunctionID, binding.Range{0, 31}).
-			Action().LoadRegRange(int(marksReg), cnpDropMark, cnpDropMarkRange).
-			Action().LoadRegRange(int(marksReg), DispositionDrop, APDispositionMarkRange). //Logging
-			Action().SendToController(uint8(PacketInReasonNP)).
-			Action().GotoTable(metricTableID).
-			Cookie(c.cookieAllocator.Request(cookie.Policy).Raw()).
-			Done()
-	} else {
-		return c.pipeline[tableID].BuildFlow(ofPriority).
-			MatchConjID(conjunctionID).
-			Action().LoadRegRange(int(CNPDropConjunctionIDReg), conjunctionID, binding.Range{0, 31}).
-			Action().LoadRegRange(int(marksReg), cnpDropMark, cnpDropMarkRange).
-			Action().GotoTable(metricTableID).
-			Cookie(c.cookieAllocator.Request(cookie.Policy).Raw()).
-			Done()
+		customReason := CustomReasonLogging
+		if disposition == DispositionRej {
+			customReason += CustomReasonReject
+		}
+		flowBuilder = flowBuilder.
+			Action().LoadRegRange(int(marksReg), disposition, APDispositionMarkRange).
+			Action().LoadRegRange(int(marksReg), uint32(customReason), CustomReasonMarkRange).
+			Action().SendToController(uint8(PacketInReasonNP))
+	} else if disposition == DispositionRej {
+		flowBuilder = flowBuilder.
+			Action().LoadRegRange(int(marksReg), CustomReasonReject, CustomReasonMarkRange). // Do the rejection.
+			Action().SendToController(uint8(PacketInReasonNP))
```

Comment:
The two lines can be moved to L1461?

Reply:
If it's `Drop` and no logging, then there won't be load custom reason and sendToController actions.

Reply:
Thanks for the explanation